nemo/collections/nlp/modules/common/text_generation_strategy.py

```diff
@@ -147,26 +155,22 @@ def end_of_generation_condition(
         returns:
             a boolean tensor indicating whether the generation should stop
         """
-        if len(end_strings) == 1 and end_strings[0] == END_OF_SEQ:
+        if (len(end_strings) == 1 and end_strings[0] == END_OF_SEQ) or not end_strings:
+            # Simple scenario: only finish on end of document token.
             return prev == eod_id
-        else:
-            tokenizer = self.model.tokenizer
-            conditions = []
-            end_tokens = set()
-            end_tokens.add(eod_id)
-            for end_string in end_strings:
-                ids_1 = tokenizer.text_to_ids(f'<extra_id_1>{end_string}')
-                ids_2 = tokenizer.text_to_ids('<extra_id_1>')
-                if len(ids_1) <= len(ids_2):
-                    continue
-                token_id = ids_1[len(ids_2) :][0]
-                end_tokens.add(token_id)
-            for p, token_item in zip(prev, tokens):
-                text = tokenizer.ids_to_text(token_item.tolist())
-                conditions.append(
-                    any([text.endswith(end_string) for end_string in end_strings] + [p.item() in end_tokens])
-                )
-            return torch.tensor(conditions, dtype=torch.bool, device=tokens.device)
+
+        end_tokens, end_strings_to_check = self._get_end_of_generation_tokens_and_strings(eod_id, end_strings)
+        assert end_tokens
+
+        is_end = torch.isin(prev, torch.tensor(list(end_tokens), dtype=prev.dtype, device=prev.device))
+
+        if end_strings_to_check:
+            # The loop below is inefficient (see warning in `_get_end_of_generation_tokens_and_strings()`)
+            for idx, token_seq in enumerate(tokens):
+                text = self.model.tokenizer.ids_to_text(token_seq.tolist())
```

Comment:
instead of de-tokenizing the whole sentence, why not just de-tokenize the last N tokens where N is the len(end_string) since at worst 1 char can be 1 token?

Reply:
I'm assuming this can be significant savings over detokenizing the entire text which can be very long

Reply:
Yeah I thought about trying something like that, but didn't do it for two reasons:
1. It's not obvious to me how many tokens we should decode, since some tokens can actually be decoded into an empty string. I was thinking I'd need to start from some "best guess" value (N in your example is a good start), then increase it if somehow it's not enough. It's definitely doable, but then point 2 below came to mind...
2. I'm not convinced there's a real use case for string-based matching. AFAIK for our models we make sure that the end of response is identified by special tokens. So making the code more complex to speed-up a use case we may not care about didn't seem worth it.

That being said, I may very well be wrong w.r.t. (2), in which case I'm happy to implement something like that. Just let me know if you're aware of situations where this is actually useful (beyond debugging).

Reply:
> It's not obvious to me how many tokens we should decode
N is a hard bound, you dont need to de-tokenize more than N

Reply:
> A whole token is never really an empty string

I learned that this is not true -- there are special tokens that get decoded into empty strings, and this is tokenizer-dependent.

Anyway, I'm happy to implement it, I just want someone to tell me that there's actually a real use case for this :)

Reply:
yes, thats specifically true for control_symbols but we shouldn't have them in the text

but its fine, we can let it be as is. can add it later in a different PR if we feel we're stopping on strings for any usecase since I currently dont see us doing that